### ISSUE TITLE
Testing Strategy

### ISSUE THREAD

**Issue by mikhael28** [fsapprenticeship/fsa-hub#68]
We have a few resources to keep in mind:

- Our serverless functions deployed on AWS Lambda
- Our EC2 Sequelize REST API to query a PostgreSQL database
- Our React Web front-end
- Our React-Native front-end

We want to make sure that we handle the top 20% of test-cases that will give us 80% of the crucial test coverage that we need.

What is our testing strategy?

**Comment by mikhael28** [comment]
@pavedwave I assigned this ticket to you because you mentioned you were looking for experience in testing & I thought this would be perfect for you. Let's focus on the Native side of it, the Web App is just a magnet to the mobile app anyway. What do you think is our highest priority item to be tested, out of the following API's that are going to be integrated into the community edition:
- Push notifications 
- Scheduling API (for events to show up) zip code/state
- User Profile (subscription information, experience, rank)
- Developer Directory
- Experience API (points, exp, tied to events, how to move up in ranks?)
- Leaderboard (tied to user profile, tied to experience)
- Subscription API (Stripe Billing?)

**Comment by pavedwave** [comment]
Thanks Michael, from top-down functional testing I'd prioritize this order  (Native being focus):

1. User Profile (subscription information, experience, rank)
2. Scheduling API (for events to show up) zip code/state
3. Push notifications
4. Experience API (points, exp, tied to events, how to move up in ranks?)
5. Developer Directory
6. Leaderboard (tied to user profile, tied to experience) 
7. Subscription API (Stripe Billing?)

My assumption is most visible end-user functionality and database integrity undergoes unit testing on each API  -- the owner of each component adding some starter unit tests, which also act as a guide for additional tests.  

Push notification functionality could be tested after event scheduling is intact, since this is where notifications matter to users most (if I'm understanding the feature correctly.)  The other option is to make sure all database-dependent features work first, and bump down Push notification testing to #6 instead.  I'd prefer this, just depends how quickly we want to publish.

Billing can come last since I'm assuming we'll rely on 3rd party billing and just need to validate the integration with a unit test?

I'll try to fast forward in my React bootcamp to the testing section for the actual coding on this! and I think I could learn a lot of React context from reviewing aj's previously written testcode.  

Thought I'd seen more testcode on this earlier -- have we archived or moved an existing suite somewhere?
https://github.com/fsapprenticeship/fsa-hub/tree/develop/react-native/ios/fsahubTests

Manual test:  With the UI in flux and review right now, beta-exploratory testing for critical OS/browser bugs is fast and efficient.   We just need to keep a record which devices and browsers we're physically testing, and fill the holes with iOS and Android emulators.

Bug Tracking:  Depending on the amount of bugs and longevity of the project, we could continue using Github Issues here, adding some conventions to make Severity and State more clear.  Later if appropriate we could also move to a tool like Bugzilla- understanding there's some admin/overhead.

**Comment by mikhael28** [comment]
@pavedwave RE the test-code, AJ's tests are in the web repository. The link with the directory tree you send me is the react-native repo, that is not expo. We are currently using the expo repo for everything. I might need to remove that to avoid confusion :)

I think organizing as much of the work on GitHub as possible is a good move, keeping issues there will also let us the 'Projects' and 'Milestones' functionality that kind of removes the need for Trello or Jira once everyone is using it. Honestly, the number of features GitHub has are mind-boggling, even though their search experience is just terrible... Bugzilla, and things like that, offer only minimal value compared to GitHub issues.

**Comment by mikhael28** [comment]
@AspenJames what do you think of @pavedwave 's plan? You are a testing wizard!
@nicholaswbowen do you have any feedback?

**Comment by pavedwave** [comment]
@mikhael28  Yeah Github's got plenty of decent tracking fields we're already using like Label, Milestone, and Open/Closed state, making it bug-trackable enough for smaller projects. Thanks for clarifying the directories and existing tests!

**Comment by mikhael28** [comment]
I'm going to make an executive decision here, and announce that we are going to use cypress.io for web applications, as well as Jest for React Web and Native. Will probably use Nock/the name I can't quite remember for testing our APIs, need to reference some of my notes.
